tests/zstreamtest.c

```diff
@@ -1777,6 +1777,91 @@ static int basicUnitTests(U32 seed, double compressibility)
+            &externalMatchState,
+            zstreamExternalMatchFinder
+        );
+
+        for (enableFallback = 0; enableFallback < 1; enableFallback++) {
+            size_t testCaseId;
+
+            EMF_testCase const EMF_successCases[] = {
+                EMF_ONE_BIG_SEQ,
+                EMF_LOTS_OF_SEQS,
+            };
+            size_t const EMF_numSuccessCases = 2;
+
+            EMF_testCase const EMF_failureCases[] = {
+                EMF_ZERO_SEQS,
+                EMF_BIG_ERROR,
+                EMF_SMALL_ERROR,
+            };
+            size_t const EMF_numFailureCases = 3;
+
+            /* Test external matchfinder success scenarios */
+            for (testCaseId = 0; testCaseId < EMF_numSuccessCases; testCaseId++) {
+                externalMatchState = EMF_successCases[testCaseId];
+                ZSTD_CCtx_reset(zc, ZSTD_reset_session_only);
+                CHECK_Z(ZSTD_CCtx_setParameter(zc, ZSTD_c_enableMatchFinderFallback, enableFallback));
+                res = ZSTD_compress2(zc, dstBuf, dstBufSize, CNBuffer, CNBufferSize);
+                CHECK(ZSTD_isError(res), "EMF: Compression error: %s", ZSTD_getErrorName(res));
+                CHECK_Z(ZSTD_decompress(checkBuf, checkBufSize, dstBuf, res));
+                CHECK(memcmp(CNBuffer, checkBuf, CNBufferSize) != 0, "EMF: Corruption!");
+            }
+
+            /* Test external matchfinder failure scenarios */
+            for (testCaseId = 0; testCaseId < EMF_numFailureCases; testCaseId++) {
+                externalMatchState = EMF_failureCases[testCaseId];
+                ZSTD_CCtx_reset(zc, ZSTD_reset_session_only);
+                CHECK_Z(ZSTD_CCtx_setParameter(zc, ZSTD_c_enableMatchFinderFallback, enableFallback));
+                res = ZSTD_compress2(zc, dstBuf, dstBufSize, CNBuffer, CNBufferSize);
+                if (enableFallback) {
+                    CHECK_Z(ZSTD_decompress(checkBuf, checkBufSize, dstBuf, res));
+                    CHECK(memcmp(CNBuffer, checkBuf, CNBufferSize) != 0, "EMF: Corruption!");
+                } else {
+                    CHECK(!ZSTD_isError(res), "EMF: Should have raised an error!");
+                    CHECK(
+                        ZSTD_getErrorCode(res) != ZSTD_error_externalMatchFinder_failed,
+                        "EMF: Wrong error code: %s", ZSTD_getErrorName(res)
+                    );
+                }
+            }
+
+            /* Test compression with external matchfinder + empty src buffer */
+            externalMatchState = EMF_ZERO_SEQS;
+            ZSTD_CCtx_reset(zc, ZSTD_reset_session_only);
+            CHECK_Z(ZSTD_CCtx_setParameter(zc, ZSTD_c_enableMatchFinderFallback, enableFallback));
+            res = ZSTD_compress2(zc, dstBuf, dstBufSize, CNBuffer, 0);
+            CHECK(ZSTD_isError(res), "EMF: Compression error: %s", ZSTD_getErrorName(res));
+            CHECK(ZSTD_decompress(checkBuf, checkBufSize, dstBuf, res) != 0, "EMF: Empty src round trip failed!");
+        }
+
```

Comment:
Should you have a test for an `RLE` block specifically ?

Reply:
On your suggestion, I implemented a [test](https://github.com/embg/zstd/commit/4cd75486cd8a7fd34c3dbb83248abb31dda79543) for RLE handling. It made me realize that we actually can’t have RLE be completely invisible to external matchfinder libraries. The test revealed that even with the mitigations in this PR + additional mitigations, an adversarial matchfinder may compress 10x worse than zstd on constant input:

```
test 75 : External matchfinder API: Error => EMF: RLE compressed size doesn't match, 5603 != 339 (seed 3173672436, test nb 76, line 1947) 
```

Context: an external match finder, for example one which inherited DEFLATE’s maximum match length, may produce a large number of sequences for an RLE block. This breaks our [heuristics](https://github.com/facebook/zstd/blob/dev/lib/compress/zstd_compress.c#L3086) for [detecting](https://github.com/facebook/zstd/blob/dev/lib/compress/zstd_compress.c#L3869) RLE blocks.

I thought simply disabling that heuristic when `useExternalMatchFinder == 1` is sufficient to handle RLE properly for any external match finder (disabling the heuristic means we always fall back to ZSTD_isRLE(), which doesn’t give false negatives).

**The problem is that zstd [will never emit the first block as RLE](https://github.com/facebook/zstd/blob/dev/lib/compress/zstd_compress.c#L3591).** So a match finder which relies entirely on zstd to post-process bad parses of constant blocks will perform really badly on the first block, even if we can post-process into RLE for subsequent blocks.

I wonder if instead of tweaking our heuristics to accommodate those bad parses, we should instead encourage match finder implementors (in zstd.h docs) to do their own RLE detection, and ensure they are producing a small number of sequences for constant blocks.

Therefore, I would like to remove anything related to RLE from this patch, except for a note in zstd.h informing implementors that they need to handle it themselves. The existing RLE mitigations in this patch would be great, *if* we didn’t have that exception about the first block of a frame :)

@Cyan4973 thoughts?

Reply:
Discussed offline -- post-processing bad parses (not just for constant blocks, but in general) is a future topic we want to work on, but it's out of scope for this PR. Post processing should also be gated by a cParam, so the application can control whether sequences are rewritten by zstd.